cmd/cloudFoundryDeploy.go

```diff
@@ -0,0 +1,855 @@
+
+	// TODO set HOME to config.DockerWorkspace
+	command.SetEnv(additionalEnvironment)
+
+	err = _cfLogin(command, cloudfoundry.LoginOptions{
+		CfAPIEndpoint: config.APIEndpoint,
+		CfOrg:         config.Org,
+		CfSpace:       config.Space,
+		Username:      config.Username,
+		Password:      config.Password,
+		CfAPIOpts:     splitAtWhitespace(config.APIParameters),
+		CfLoginOpts:   splitAtWhitespace(config.LoginParameters),
+	})
+
+	if err == nil {
+		loginPerformed = true
+	}
+
+	if err == nil {
+		err = command.RunExecutable("cf", []string{"plugins"}...)
+		if err != nil {
+			log.Entry().WithError(err).Errorf("Command '%s' failed.", []string{"plugins"})
+		}
+	}
+
+	if err == nil {
+		err = command.RunExecutable("cf", cfDeployParams...)
+		if err != nil {
+			log.Entry().WithError(err).Errorf("Command '%s' failed.", cfDeployParams)
+		}
+	}
+
+	if err == nil && postDeployAction != nil {
+		err = postDeployAction(command)
+	}
+
+	if loginPerformed {
+
+		logoutErr := _cfLogout(command)
+
+		if logoutErr != nil {
+			log.Entry().WithError(logoutErr).Errorf("Cannot perform cf logout")
+			if err == nil {
+				err = logoutErr
+			}
+		}
+	}
+
+	if err != nil || GeneralConfig.Verbose {
+		e := handleCfCliLog(cfLogFile)
+		if e != nil {
+			log.Entry().WithError(err).Errorf("Error reading cf log file '%s'.", cfLogFile)
+		}
+	}
+
+	return err
+}
+
+func findMtar() (string, error) {
+
+	const pattern = "**/*.mtar"
+
+	mtars, err := fileUtils.Glob(pattern)
+
+	if err != nil {
+		return "", err
+	}
+
+	if len(mtars) == 0 {
+		return "", fmt.Errorf("No mtar file matching pattern '%s' found", pattern)
+	}
+
+	if len(mtars) > 1 {
+		sMtars := []string{}
+		for _, mtar := range mtars {
+			sMtars = append(sMtars, mtar)
+		}
+		return "", fmt.Errorf("Found multiple mtar files matching pattern '%s' (%s), please specify file via mtaPath parameter 'mtarPath'", pattern, strings.Join(sMtars, ","))
+	}
+
+	return mtars[0], nil
+}
+
+func handleCfCliLog(logFile string) error {
+
+	fExists, err := fileUtils.FileExists(logFile)
+
+	if err != nil {
+		return err
+	}
+
+	log.Entry().Info("### START OF CF CLI TRACE OUTPUT ###")
+
+	if fExists {
+
+		f, err := os.Open(logFile)
+
+		if err != nil {
+			return err
+		}
+
+		bReader := bufio.NewReader(f)
+
+		var done bool
+
+		for {
+			line, err := bReader.ReadString('\n')
+			if err == io.EOF {
+
+				// maybe inappropriate to log as info. Maybe the line from the
+				// log indicates an error, but that is something like a project
+				// standard.
+				done = true
+			} else if err != nil {
+
+				break
+			}
+
+			log.Entry().Info(strings.TrimSuffix(line, "\n"))
+
+			if done {
+				break
+			}
+		}
+
+	} else {
+		log.Entry().Warningf("No trace file found at '%s'", logFile)
+	}
+
+	log.Entry().Info("### END OF CF CLI TRACE OUTPUT ###")
+
+	return err
+}
+
+func splitAtWhitespace(s string) []string {
+	if len(strings.TrimSpace(s)) == 0 {
+		return []string{}
+	}
+	return regexp.MustCompile("\\s+").Split(s, -1)
```

Comment:
Just for my education: Is this different from return string.Split(s, " ")?

Reply:
`strings.Split(./.)` preserves the blanks in the result. But we don't want to have the blanks when preparing string slices which are used later on as a command line. Beside that `strings.Split(xxxx, " ")` does also not split at e.g. tabs (Ok that is not very likely, but you know: you never know ...)

To be more precise:

```
package main
  
import (
        "fmt"
        "strings"
        "regexp"
)


func main() {

        cmd := "echo abc   123\txx"
        s := strings.Split(cmd, " ")


        for i, _s := range s {
                fmt.Printf("%d: '%s'\n", i, _s)
        }


        s = regexp.MustCompile("\\s+").Split(cmd, -1)

        for i, _s := range s {
                fmt.Printf("%d: '%s'\n", i, _s)
        }
}
```

Reply:
If your string has trailing whitespace, it will be returned in the array. If you use just `strings.Fields()`, it will leave no trailing empty entry. I think you could use `strings.Fields()` in place of the entire function.

Reply:
Didn't know that. The approach with `strings.Fields(./.)` is much simpler --> adjusted.

Reply:
I actually learned of the existence of strings.Fields() via your PR (this one here), by your use of `strings.FieldsFunc()`. ;-)